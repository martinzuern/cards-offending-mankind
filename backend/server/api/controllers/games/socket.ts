import socketIo from 'socket.io';
import socketioJwt from 'socketio-jwt';

import GameService from '../../services/game.service';
import DBService from '../../services/db.service';
import L from '../../../common/logger';
import { PlayerJwt } from '../../../../root-types';

const JWT_SECRET = process.env.JWT_SECRET;

type JwtAuthenticatedSocket = SocketIO.Socket & {
  decoded_token: PlayerJwt;
};

export default function sockets(io: socketIo.Server): void {
  io.on(
    'connection',
    socketioJwt.authorize({
      secret: JWT_SECRET,
      // eslint-disable-next-line @typescript-eslint/camelcase
      additional_auth: async (decoded: PlayerJwt, onSuccess, onError) => {
        L.info(`Player ${decoded.id} wants to connect to game ${decoded.gameId}.`);
        const canConnect = await DBService.setUserLock(decoded.id, true);

        if (canConnect) {
          return onSuccess();
        } else {
          L.warn(`Player ${decoded.id} is locked.`);
          return onError('user-locked');
        }
      },
      timeout: 1000 * 15, // 15 seconds
    })
  )
    .on('authenticated', async (socket: JwtAuthenticatedSocket) => {
      const { id: playerId, gameId } = socket.decoded_token;
      const roomName = `game:${gameId}`;
      L.info(`Player ${playerId} authenticated for ${roomName}.`);

      const gameState = await DBService.getGame(gameId);
      if (!GameService.isGameJoinable(gameState.game)) {
        L.warn(`Game ${gameId} is not joinable.`);
        socket.emit('exception', { errorMessage: 'Game is not joinable.' });
        socket.disconnect(true);
        return;
      }

      // make sure we keep the user lock
      socket.conn.on('packet', async (packet) => {
        if (packet.type === 'ping') {
          await DBService.setUserLock(playerId);
        }
      });

      // Add to correct room
      socket.join(roomName);
    })
    .on('disconnect', async (socket: JwtAuthenticatedSocket) => {
      const playerId = socket?.decoded_token?.id;
      if (!playerId) return;
      await DBService.deleteUserLock(playerId);
    });
}
