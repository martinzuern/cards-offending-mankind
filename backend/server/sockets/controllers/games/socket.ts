import socketIo from 'socket.io';
import socketioJwt from 'socketio-jwt';
import _ from 'lodash';

import DBService from '../../../services/db.service';
import Controller, { JwtAuthenticatedSocket } from './controller';
import wrapAsync from '../../middlewares/error.handler';
import L from '../../../common/logger';

const { JWT_SECRET } = process.env;

const auth = socketioJwt.authorize({ secret: JWT_SECRET, additional_auth: Controller.validateJwt });

export default function sockets(io: socketIo.Server): void {
  io.on('connection', auth)
    .on('disconnect', Controller.onDisconnect)
    .on('authenticated', async (socket: JwtAuthenticatedSocket) => {
      const { id: playerId, gameId } = socket.decoded_token;
      const room = `game:${gameId}`;

      // make sure we keep the user lock
      socket.conn.on('packet', async (packet) => {
        if (packet.type === 'ping') {
          await DBService.setUserLock(playerId);
        }
      });

      L.info(`Player ${playerId} authenticated for game ${gameId}.`);
      socket.join(room);

      const c = new Controller(io, socket, room);
      const wrapper = _.partial(wrapAsync, socket);

      wrapper(c.joinGame)();
      socket.on('start-game', wrapper(c.startGame));
    });
}
