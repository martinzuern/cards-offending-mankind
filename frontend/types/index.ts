/// <reference path="../../types/index.ts" />

export * from '../../types';