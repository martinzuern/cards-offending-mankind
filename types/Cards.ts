export interface Pack {
  abbr: string;
  name: string;
  icon: string;
  description: string;
  official: boolean;
}

export interface PackInformation extends Pack {
  promptsCount: number;
  responsesCount: number;
}

export enum CardType {
  // Currently, we only support text cards
  Text = 'text',
}

interface GenericCard {
  type: CardType;
  value: string;
  packAbbr: Pack['abbr'];
}

export interface PromptCard extends GenericCard {
  draw: number;
  pick: number;
}

export interface ResponseCard extends GenericCard {}
