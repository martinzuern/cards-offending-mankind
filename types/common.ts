export type UUID = string & { readonly _: unique symbol };
