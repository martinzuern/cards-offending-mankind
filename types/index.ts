export * from './common';
export * from './Game';
export * from './Cards';
export * from './Player';
