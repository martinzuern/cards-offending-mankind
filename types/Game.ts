import { UUID } from './common';
import { Pack, PromptCard, ResponseCard } from './Cards';
import { OtherPlayer, Player } from './Player';

export enum GameStatus {
  Created = 'created',
  Running = 'running',
  Ended = 'ended',
}

export interface Game {
  id: UUID;
  hasPassword: boolean;
  status: GameStatus;
  packs: Pack[];
  handSize: number;
  winnerPoints: number | false;
  timeouts: {
    playing: number;
    revealing: number;
    judging: number;
    betweenRounds: number;
  };
  specialRules: {
    aiPlayerCount: 0 | 1 | 2 | 3;
    allowDiscarding: {
      enabled: boolean;
      penalty: number;
    };
  };
}

export interface CreateGame extends Partial<Omit<Game, 'id' | 'status' | 'hasPassword' | 'packs'>> {
  password?: string;
  packs: (Pack | Pick<Pack, 'abbr'>)[];
}

export enum RoundStatus {
  Created = 'created',
  Played = 'played',
  Revealed = 'revealed',
  Ended = 'ended',
}

type timeouts = {
  playing: Date;
  revealing: Date;
  judging: Date;
  betweenRounds: Date;
};

export type RoundTimeoutKeys = keyof timeouts;

export interface RoundSubmission {
  playerId: Player['id'];
  timestamp: Date;
  cards: ResponseCard[];
  pointsChange: number;
  isRevealed: boolean;
}

export interface Round {
  judgeId: Player['id'];
  status: RoundStatus;
  timeouts:
    | Pick<timeouts, 'playing'>
    | Pick<timeouts, 'playing' | 'revealing'>
    | Pick<timeouts, 'playing' | 'revealing' | 'judging'>
    | timeouts;
  prompt: PromptCard;
  submissions: RoundSubmission[];
  discard: RoundSubmission[];
}

export interface GameState {
  game: Game;
  players: OtherPlayer[];
  rounds: Round[];
}
