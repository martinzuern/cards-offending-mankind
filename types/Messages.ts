import { ResponseCard } from './Cards';
import { Round, Game, CreateGame } from './Game';
import { PlayerWithToken, CreatePlayer } from '.';

export interface MessageCreateGame {
  game: CreateGame;
  player: CreatePlayer;
}
export interface MessageGameCreated {
  game: Game;
  player: PlayerWithToken;
}

export interface MessageJoinGame extends CreatePlayer {
  password?: string;
}
export interface MessagePlayerJoined {
  player: PlayerWithToken;
}
export interface MessageGetGame {
  game: Game;
}

export interface MessagePickCards {
  roundIndex: number;
  cards: ResponseCard[];
}

export interface MessageDiscardPrompt {
  roundIndex: number;
}

export interface MessageRevealSubmission {
  roundIndex: number;
  submissionIndex: number;
}

export interface MessageChooseWinner {
  roundIndex: number;
  submissionIndex: number;
}

export interface MessageRoundUpdated {
  roundIndex: number;
  round: Round;
}
