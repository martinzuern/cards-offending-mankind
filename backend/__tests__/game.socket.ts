import { Socket, io as ioFront } from 'socket.io-client';
import { AddressInfo } from 'net';
import http from 'http';
import fetch from 'node-fetch';
import assert from 'assert';
import { v4 as uuidv4 } from 'uuid';

import Server from '../server';
import { GameState, Player } from '../../types';

let httpServer: http.Server;
let Teardown: () => Promise<void>;
let socket: Socket;
let httpServerUrl: URL;

const promiseTimeout = (ms, promise) => {
  // Create a promise that rejects in <ms> milliseconds
  const timeout = new Promise((_, reject) => {
    let id = setTimeout(() => {
      clearTimeout(id);
      reject('Timed out in ' + ms + 'ms.');
    }, ms);
  });

  return Promise.race([promise, timeout]);
};

const postRequest = async (path, data): Promise<unknown> => {
  const res = await fetch(new URL(path, httpServerUrl), {
    method: 'post',
    body: JSON.stringify(data),
    headers: { 'Content-Type': 'application/json' },
  });
  const json = await res.json();
  assert(res.ok, `Error â€“ ${json}`);
  return json;
};

const newSocket = async (token: string): Promise<Socket> => {
  const s = ioFront(httpServerUrl.toString(), {
    extraHeaders: { Authorization: `Bearer ${token}` },
    reconnection: false,
    forceNew: true,
  });
  return promiseTimeout(
    500,
    new Promise<Socket>((resolve) => {
      s.on('connect', () => {
        expect(s.connected).toBeTruthy();
        resolve(s);
      });
    })
  );
};

const createNewGame = async (data: {}): Promise<GameState> =>
  (postRequest('api/v1/games', {
    player: { nickname: 'foo' },
    game: { ...data },
  }) as unknown) as GameState;

const joinGame = async (createdGame: GameState, nickname = '') => {
  nickname = nickname || uuidv4();
  return postRequest(`/api/v1/games/${createdGame.game.id}/join`, {
    nickname,
  });
};

beforeAll(async (done) => {
  const { server, closeServer } = await Server;
  httpServer = server;
  Teardown = closeServer;

  httpServer.listen(undefined, '127.0.0.1', () => {
    const addr = httpServer.address() as AddressInfo;
    httpServerUrl = new URL(`http://localhost:${addr.port}`);
    done();
  });
});

afterAll((done) => Teardown().then(done), 90000);

afterEach(() => {
  if (socket && socket.connected) socket.disconnect();
});

describe('joining game', () => {
  let createdGame;
  beforeAll(async (done) => {
    createdGame = await createNewGame({});
    done();
  });

  it('error jwt malformed', async (done) => {
    let token = 'this-aint-work';
    socket = ioFront(httpServerUrl.toString(), {
      extraHeaders: { Authorization: `Bearer ${token}` },
      reconnection: false,
      forceNew: true,
    });
    socket.on('connect_error', (error) => {
      expect(error).toMatchSnapshot();
      done();
    });
  });

  it('error jwt not valid', (done) => {
    let token =
      'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6ImFmZGIxNzE4LWIyOTAtNDY4Yy1hZjc2LTk5MWE0ZTU5NGEwYiIsImdhbWVJZCI6ImM2NTZlOTUxLWQzMDQtNDZhYS1hZDE3LWIwN2MyMzk4NWQxNiIsImlhdCI6MTU4OTkyMzA1MSwiZXhwIjoxNTkwMDA5NDUxfQ.R-5Q0kACeySzySOh3rFyyzM43JoghbTaeHijy9X4DUI';
    socket = ioFront(httpServerUrl.toString(), {
      extraHeaders: { Authorization: `Bearer ${token}` },
      reconnection: false,
      forceNew: true,
    });
    socket.on('connect_error', (error) => {
      expect(error).toMatchSnapshot();
      done();
    });
  });

  it('accepts valid token', async (done) => {
    socket = await newSocket(createdGame.player.token);
    socket
      .on('authenticated', () => {})
      .on('gamestate_updated', (data) => {
        expect(data).toMatchSnapshot({
          game: {
            id: expect.any(String),
          },
          players: [
            {
              id: expect.any(String),
            },
          ],
        });
        done();
      });
  });

  it('error on locked user', async (done) => {
    socket = await newSocket(createdGame.player.token);
    const socket2 = ioFront(httpServerUrl.toString(), {
      extraHeaders: { Authorization: `Bearer ${createdGame.player.token}` },
      reconnection: false,
      forceNew: true,
    });
    socket2.on('connect_error', (error) => {
      expect(error).toMatchSnapshot();
      socket2.disconnect();
      done();
    });
  });
});

describe('error starting game', () => {
  let createdGame;
  beforeEach(async (done) => {
    createdGame = await createNewGame({});
    done();
  });

  it('error when there are not enough players', async (done) => {
    socket = await newSocket(createdGame.player.token);
    socket.emit('start_game');
    socket
      .on('exception', (error) => {
        expect(error).toMatchSnapshot();
        done();
      })
      .on('connect_error', (error) => {
        throw new Error(JSON.stringify(error));
      });
  });

  describe('with two players', () => {
    let socket2;
    let player2;
    let socket3;
    let player3;
    beforeEach(async (done) => {
      socket = await newSocket(createdGame.player.token);
      player2 = await joinGame(createdGame);
      player3 = await joinGame(createdGame);
      socket2 = await newSocket(player2.player.token);
      socket3 = await newSocket(player3.player.token);
      done();
    });

    afterEach(() => {
      if (socket2.connected) socket2.disconnect();
      if (socket3.connected) socket3.disconnect();
    });

    it('error when non-host starts game', (done) => {
      socket2.emit('start_game');
      socket2
        .on('exception', (error) => {
          expect(error).toMatchSnapshot();
          done();
        })
        .on('connect_error', (error) => {
          throw new Error(JSON.stringify(error));
        });
    });

    it('error when there are not enough packs', async (done) => {
      socket.emit('start_game');
      socket
        .on('exception', (error) => {
          expect(error).toMatchSnapshot();
          done();
        })
        .on('connect_error', (error) => {
          throw new Error(JSON.stringify(error));
        });
    });
  });
});

describe('perform game', () => {
  let createdGame;
  let socket2;
  let player2;
  let socket3;
  let player3;

  beforeEach(async (done) => {
    createdGame = await createNewGame({ packs: [{ abbr: 'Base-INTL' }] });
    done();
  });

  beforeEach(async (done) => {
    player2 = await joinGame(createdGame, 'foo2');
    player3 = await joinGame(createdGame, 'foo3');
    socket2 = await newSocket(player2.player.token);
    socket3 = await newSocket(player3.player.token);
    done();
  });

  afterEach(() => {
    if (socket2.connected) socket2.disconnect();
    if (socket3.connected) socket3.disconnect();
  });

  it('works', async (done) => {
    socket = await newSocket(createdGame.player.token);
    let latestGameState: GameState = {} as GameState;
    let latestPlayer: Player = {} as Player;
    socket
      .on('connect_error', (error) => {
        throw new Error(JSON.stringify(error));
      })
      .on('exception', (error) => {
        throw new Error(JSON.stringify(error));
      })
      .on('gamestate_updated', (data) => {
        latestGameState = data;
      })
      .on('player_updated', (data) => {
        latestPlayer = data;
      })
      .emit('start_game');

    setTimeout(() => {
      expect(latestGameState).toMatchSnapshot({
        game: {
          id: expect.any(String),
        },
        players: expect.any(Array),
        rounds: [
          {
            judgeId: expect.any(String),
            prompt: expect.any(Object),
            timeouts: {
              playing: expect.any(String),
            },
          },
        ],
      });
      expect(latestPlayer).toMatchSnapshot({
        id: expect.any(String),
        deck: expect.any(Array),
      });
      expect(
        latestPlayer.deck.length >= 10 || latestGameState.rounds[-1].judgeId === latestPlayer.id
      ).toBeTruthy();
      done();
    }, 5000);
  }, 30000);
});
