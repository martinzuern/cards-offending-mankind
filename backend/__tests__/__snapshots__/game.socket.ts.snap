// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error starting game error when there are not enough players 1`] = `
Object {
  "message": "There are not enough human players.",
  "type": "AssertionError",
}
`;

exports[`error starting game with two players error when non-host starts game 1`] = `
Object {
  "message": "Only a host can start a game.",
  "type": "AssertionError",
}
`;

exports[`error starting game with two players error when there are not enough packs 1`] = `
Object {
  "message": "There are not enough packs.",
  "type": "AssertionError",
}
`;

exports[`joining game accepts valid token 1`] = `
Object {
  "game": Object {
    "handSize": 10,
    "hasPassword": false,
    "id": Any<String>,
    "packs": Array [],
    "specialRules": Object {
      "aiPlayerCount": 0,
      "allowDiscarding": Object {
        "enabled": false,
        "penalty": 0,
      },
      "pickExtra": false,
    },
    "status": "created",
    "timeouts": Object {
      "betweenRounds": 30,
      "judging": 120,
      "playing": 120,
      "revealing": 60,
    },
    "winnerPoints": 20,
  },
  "players": Array [
    Object {
      "id": Any<String>,
      "isAI": false,
      "isActive": true,
      "isHost": true,
      "nickname": "foo",
      "points": 0,
    },
  ],
  "rounds": Array [],
}
`;

exports[`joining game error jwt malformed 1`] = `
Object {
  "data": Object {
    "code": "invalid_token",
    "message": "jwt malformed",
    "type": "UnauthorizedError",
  },
  "inner": Object {
    "message": "jwt malformed",
  },
  "message": "jwt malformed",
}
`;

exports[`joining game error jwt not valid 1`] = `
Object {
  "data": Object {
    "code": "invalid_token",
    "message": "invalid signature",
    "type": "UnauthorizedError",
  },
  "inner": Object {
    "message": "invalid signature",
  },
  "message": "invalid signature",
}
`;

exports[`joining game error on locked user 1`] = `
Object {
  "data": Object {
    "code": "user-locked",
    "message": "user-locked",
    "type": "UnauthorizedError",
  },
  "inner": Object {
    "message": "user-locked",
  },
  "message": "user-locked",
}
`;

exports[`perform game works 1`] = `
Object {
  "game": Object {
    "handSize": 10,
    "hasPassword": false,
    "id": Any<String>,
    "packs": Array [
      Object {
        "abbr": "10",
        "name": "CAH Base Set",
        "official": true,
      },
    ],
    "specialRules": Object {
      "aiPlayerCount": 0,
      "allowDiscarding": Object {
        "enabled": false,
        "penalty": 0,
      },
      "pickExtra": false,
    },
    "status": "created",
    "timeouts": Object {
      "betweenRounds": 30,
      "judging": 120,
      "playing": 120,
      "revealing": 60,
    },
    "winnerPoints": 20,
  },
  "players": Array [
    Object {
      "id": Any<String>,
      "isAI": false,
      "isActive": true,
      "isHost": true,
      "nickname": "foo",
      "points": 0,
    },
    Object {
      "id": Any<String>,
      "isAI": false,
      "isActive": true,
      "isHost": false,
      "nickname": "foo2",
      "points": 0,
    },
    Object {
      "id": Any<String>,
      "isAI": false,
      "isActive": true,
      "isHost": false,
      "nickname": "foo3",
      "points": 0,
    },
  ],
  "rounds": Array [],
}
`;

exports[`perform game works 2`] = `
Object {
  "deck": Array [],
  "id": Any<String>,
  "isAI": false,
  "isActive": true,
  "isHost": true,
  "nickname": "foo",
  "points": 0,
}
`;

exports[`perform game works 3`] = `
Object {
  "game": Object {
    "handSize": 10,
    "hasPassword": false,
    "id": Any<String>,
    "packs": Array [
      Object {
        "abbr": "10",
        "name": "CAH Base Set",
        "official": true,
      },
    ],
    "specialRules": Object {
      "aiPlayerCount": 0,
      "allowDiscarding": Object {
        "enabled": false,
        "penalty": 0,
      },
      "pickExtra": false,
    },
    "status": "running",
    "timeouts": Object {
      "betweenRounds": 30,
      "judging": 120,
      "playing": 120,
      "revealing": 60,
    },
    "winnerPoints": 20,
  },
  "players": Any<Array>,
  "rounds": Array [
    Object {
      "discard": Array [],
      "judgeId": Any<String>,
      "prompt": Any<Object>,
      "status": "created",
      "submissions": Array [],
      "timeouts": Object {
        "playing": Any<String>,
      },
    },
  ],
}
`;

exports[`perform game works 4`] = `
Object {
  "deck": Any<Array>,
  "id": Any<String>,
  "isAI": false,
  "isActive": true,
  "isHost": true,
  "nickname": "foo",
  "points": 0,
}
`;
